### PR TITLE
Honour sticky header height while scrolling in RLV

### PR DESCRIPTION
**Pull request by swapnil1104**
## Issues faced
 - The StickyObject currently does not honor the height of sticky header while updating **smallestVisibleIndex** field as it is directly dependent on onVisibleIndicesChanged callback from RecyclerListView.
 - Cannot pass style props to the StickyObject container.

## Proposed changes
 - Replacing distanceFromWindow with **scrollOffsetCorrectionDelta** in StickyContainer for more generalized application.
 - Addition of **stickyContainerRenderer** to StickyContainer which allows devs to provide their own implementation of container wrapper for the contents of StickyObjects.

### CODE REVIEW

File: src/core/sticky/StickyFooter.tsx [hunk 1]
```diff
@@ -50,9 +50,9 @@ export default class StickyFooter<P extends StickyObjectProps> extends StickyObj
         return scrollableHeight ? -1 * (offsetY + scrollableHeight) : undefined;
     }
 
-    protected hasReachedBoundary(offsetY: number, distanceFromWindow: number, windowBound?: number): boolean {
+    protected hasReachedBoundary(offsetY: number, windowBound?: number): boolean {
         if (windowBound) {
-            const endReachedMargin = Math.round(offsetY - (windowBound + distanceFromWindow));
+            const endReachedMargin = Math.round(offsetY - (windowBound));
```
Comment (anshulika-prasad) [inline src/core/sticky/StickyFooter.tsx]: Please verify if removing `distanceFromWindow` works fine

File: src/core/sticky/StickyHeader.tsx [hunk 2]
```diff
@@ -52,7 +51,7 @@ export default class StickyHeader<P extends StickyObjectProps> extends StickyObj
         return offsetY;
     }
 
-    protected hasReachedBoundary(offsetY: number, distanceFromWindow: number, _windowBound?: number): boolean {
-        return offsetY <= distanceFromWindow;
+    protected hasReachedBoundary(offsetY: number, _windowBound?: number): boolean {
+        return false;
```
Comment (anshulika-prasad) [inline src/core/sticky/StickyHeader.tsx]: Please explain ??

File: src/core/sticky/StickyHeader.tsx [hunk 3]
```diff
@@ -52,7 +51,8 @@ export default class StickyHeader<P extends StickyObjectProps> extends StickyObj
         return offsetY;
     }
 
-    protected hasReachedBoundary(offsetY: number, distanceFromWindow: number, _windowBound?: number): boolean {
-        return offsetY <= distanceFromWindow;
+    protected hasReachedBoundary(offsetY: number, _windowBound?: number): boolean {
```
Comment (swapnil1104) [inline src/core/sticky/StickyHeader.tsx]: @ananyachandra14 Please review this logic.

File: src/core/RecyclerListView.tsx [hunk 4]
```diff
@@ -579,11 +579,11 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
...
         //Adjusting offsets using distanceFromWindow
-        let scrollCorrection = 0;
-        if (this.props.correctedScrollOffset) {
-            scrollCorrection = this.props.correctedScrollOffset();
+        let correctionDelta = 0;
+        if (this.props.scrollOffsetCorrectionDelta) {
+            correctionDelta = this.props.scrollOffsetCorrectionDelta();
         }
-        this._virtualRenderer.updateOffset(offsetX, offsetY, -scrollCorrection, true);
+        this._virtualRenderer.updateOffset(offsetX, offsetY, correctionDelta, true);
```
Comment (muskeinsingh) [inline src/core/RecyclerListView.tsx]: `-` missing?

Comment (muskeinsingh) [reply src/core/RecyclerListView.tsx]: If not, can you update with a comment why this call was taken?

Comment (swapnil1104) [reply src/core/RecyclerListView.tsx]: Added

File: docs/guides/sticky/README.md [hunk 5]
```diff
@@ -17,7 +17,10 @@ _setRef(recycler) {
...
-* `style`                 - Optional. Pass the same style that is applied to the RecyclerListView component here.
+* `stickyHeaderIndices`     - An array of indices whose corresponding items need to be stuck to the top of the RecyclerListView once the items scroll off the top. Every subsequent sticky index view will push the previous sticky view off the top to take its place. Needs to be sorted ascending.
+* `stickyFooterIndices`     - Works same as sticky headers, but for views to be stuck at the bottom of the recyclerView. Needs to be sorted ascending.
+* `overrideRowRenderer`     - Optional. Will be called instead of rowRenderer for all sticky items. Any changes to the item for when they are stuck can be done here. Refer to sample code for usage.
+* `stickyContainerRenderer` - Optional. Pass a stylized container for StickyHeader and StickyFooter, providing user extensibility to customize the look and feel of these items.
+* `scrollOffsetCorrection`  - Optional. Will be called upon onScroll of the RecyclerListView, and scroll offset will be corrected based on the value returned from this function.
```
Comment (naqvitalha) [inline docs/guides/sticky/README.md]: Name doesn't sound right. You should call it getLogicalOffset and define logical offset and it's impact and how it is different from actual opne

File: src/core/RecyclerListView.tsx (hunk 4 above)
Comment (naqvitalha) [reply src/core/RecyclerListView.tsx]: From contract we should take the entire new offset externally

File: src/core/StickyContainer.tsx [hunk 6]
```diff
@@ -23,6 +23,8 @@ export interface StickyContainerProps {
     stickyHeaderIndices?: number[];
     stickyFooterIndices?: number[];
     overrideRowRenderer?: (type: string | number | undefined, data: any, index: number, extendedState?: object) => JSX.Element | JSX.Element[] | null;
+    scrollOffsetCorrection?: () => number;
+    stickyContainerRenderer?: (stickyContent: JSX.Element, index: number, extendedState?: object) => JSX.Element | null;
```
Comment (naqvitalha) [inline src/core/StickyContainer.tsx]: shouldn't there be a props section?

Comment (swapnil1104) [reply src/core/StickyContainer.tsx]: Added the new params to propTypes, please refer to lines 238 - 243.

File: src/core/RecyclerListView.tsx [hunk 7]
```diff
@@ -583,8 +587,12 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
     }
 
     private _onScroll = (offsetX: number, offsetY: number, rawEvent: ScrollEvent): void => {
-        //Adjusting offsets using distanceFromWindow
-        this._virtualRenderer.updateOffset(offsetX, offsetY, -this.props.distanceFromWindow!, true);
+        // correction to be positive to shift offset upwards; negative to push offset downwards.
+        // extracting the correction value from logical offset and updating offset of virtual renderer.
+        if (this.props.getWindowCorrection) {
+            this._windowCorrection =  this.props.getWindowCorrection();
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: Pass object to avoid object creation

Comment (naqvitalha) [reply src/core/RecyclerListView.tsx]: *same object

File: src/core/ViewabilityTracker.ts [hunk 8]
```diff
@@ -73,17 +82,22 @@ export default class ViewabilityTracker {
 
     public forceRefreshWithOffset(offset: number): void {
```
Comment (naqvitalha) [inline src/core/ViewabilityTracker.ts]: accept correct here as well

File: src/core/StickyContainer.tsx [hunk 9]
```diff
@@ -206,8 +213,14 @@ export default class StickyContainer<P extends StickyContainerProps> extends Com
         return undefined;
     }
 
-    private _getDistanceFromWindow = (): number => {
-        return this._distanceFromWindow;
+    private _applyWindowCorrection = (offset: number, windowCorrection: WindowCorrection): void => {
+        if (this.props.applyWindowCorrection) {
+            this.props.applyWindowCorrection(offset, windowCorrection);
+            this._windowCorrection = windowCorrection;
```
Comment (naqvitalha) [inline src/core/StickyContainer.tsx]: Not required

File: src/core/StickyContainer.tsx [hunk 10]
```diff
@@ -206,8 +213,14 @@ export default class StickyContainer<P extends StickyContainerProps> extends Com
...
+        }
+        if (this._stickyHeaderRef && this._stickyHeaderRef.layoutRect) {
+            this._windowCorrection.startCorrection += Math.ceil(this._stickyHeaderRef.layoutRect.height);
```
Comment (naqvitalha) [inline src/core/StickyContainer.tsx]: Will lead to inconsistency, don't modify this internally

File: src/core/RecyclerListView.tsx [hunk 11]
```diff
@@ -484,7 +488,7 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
             this._pendingScrollToOffset = offset;
             this.setState({});
         } else {
-            this._virtualRenderer.startViewabilityTracker();
+            this._virtualRenderer.startViewabilityTracker(this._windowCorrection);
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: Should we ever use this value directly?

File: src/core/RecyclerListView.tsx [hunk 12]
```diff
@@ -583,8 +587,12 @@ export default class RecyclerListView<P extends RecyclerListViewProps, S extends
     }
 
     private _onScroll = (offsetX: number, offsetY: number, rawEvent: ScrollEvent): void => {
-        //Adjusting offsets using distanceFromWindow
-        this._virtualRenderer.updateOffset(offsetX, offsetY, -this.props.distanceFromWindow!, true);
+        // correction to be positive to shift offset upwards; negative to push offset downwards.
+        // extracting the correction value from logical offset and updating offset of virtual renderer.
+        if (this.props.applyWindowCorrection) {
+            this.props.applyWindowCorrection(this.props.isHorizontal ? offsetX : offsetY, this._windowCorrection);
```
Comment (naqvitalha) [inline src/core/RecyclerListView.tsx]: why not pass both?

File: src/core/ViewabilityTracker.ts [hunk 13]
```diff
@@ -289,12 +301,16 @@ export default class ViewabilityTracker {
         return this._itemIntersectsWindow(this._visibleWindow, startBound, endBound);
     }
 
-    private _updateTrackingWindows(newOffset: number): void {
-        this._engagedWindow.start = Math.max(0, newOffset - this._renderAheadOffset);
-        this._engagedWindow.end = newOffset + this._windowBound + this._renderAheadOffset;
+    private _updateTrackingWindows(offset: number, correctedOffset: number, correction: WindowCorrection): void {
```
Comment (naqvitalha) [inline src/core/ViewabilityTracker.ts]: Why do we need offset here?

File: src/core/sticky/StickyObject.tsx [hunk 14]
```diff
@@ -66,41 +71,46 @@ export default abstract class StickyObject<P extends StickyObjectProps> extends
...
-                {position: "absolute", width: this._scrollableWidth, transform: [{translateY: this._stickyViewOffset}]},
-                this.containerPosition,
-            ]}>
-                {this.stickyVisiblity ?
-                    this._renderSticky()
-                : null}
+        // Add the container style if renderContainer is undefined
+        const content =  (
+            <Animated.View onLayout={this._onLayout}
+                style={[{transform : [{translateY: this._stickyViewOffset}]}, (this.props.renderContainer === undefined ? this._containerStyle : null)]}>
```
Comment (naqvitalha) [inline src/core/sticky/StickyObject.tsx]: This is wrong, why not construct the object here?

File: src/core/sticky/StickyObject.tsx [hunk 15]
```diff
@@ -66,41 +71,46 @@ export default abstract class StickyObject<P extends StickyObjectProps> extends
...
+            return (content);
+        }
     }
 
     public onVisibleIndicesChanged(all: number[]): void {
         if (this._firstCompute) {
             this.initStickyParams();
             this._firstCompute = false;
         }
+        this._containerStyle = [{ position: "absolute", width: this._scrollableWidth }, this.containerPosition];
```
Comment (naqvitalha) [inline src/core/sticky/StickyObject.tsx]: This needs to be relooked

**Pull request merged**